Fix file paths for uncached transformed images

RJacksonm1 opened the pull request:
When resizing an image the `sourceFilename` (and thus its `.Contents`) represented the original file _if_ the resized image was not already within Hugo's resource cache. If the resized file was already in the resource cache, `sourceFilename` (and `.Contents`) were set as expected.

This PR defers setting `sourceFilename` until we already have an image, whether it was loaded from cache or created anew.

Fixes #5012

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/gohugoio/hugo?pullRequest=5013) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/gohugoio/hugo?pullRequest=5013) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/gohugoio/hugo?pullRequest=5013) it.</sub>

bep commented:
@RJacksonm1 I suspect that the test that fails the CI build is actually a test for this particular issue, but where the assertion is wrong.

RJacksonm1 commented:
@bep Ah cheers. I'm having trouble getting my tests running at the minute (`use of internal package not allowed`, despite having my code in `/Users/rjackson/Go/src/github.com/gohugoio/hugo` 🤷‍♂️ ).

Once I get the tests working, I'll have a crack at making sure this issue is properly covered.

bep commented:
>use of internal package not allowed

Hmm... I've been in that boat for other projects. A vendoring issue? If you use `mage vendor` from the `hugo` folder it should work.

RJacksonm1 commented:
@bep No luck with that alone. Looks like I've solved it by nuking everything and re-downloading the source as per the [contributing guide](https://github.com/gohugoio/hugo/blob/master/CONTRIBUTING.md#fetching-the-sources-from-github). I was previously following the README's instructions (though I'd imagine the fault is more likely on my side regardless)

I now have the failing `TestImageTransformBasic` test 🙌

RJacksonm1 commented:
Alright, tests updated.

I've rebased the change so my new test for the correct `sourceFilename` behaviour runs before updating the code, so we can compare before/after the fix:

```
git rebase --exec 'go test github.com/gohugoio/hugo/resource -run "^TestImageTransformBasic$" || true' origin/master

Executing: go test github.com/gohugoio/hugo/resource -run "^TestImageTransformBasic$" || true
     sunset.jpg	en	Filename: content/sunset.jpg	Base: content
--- FAIL: TestImageTransformBasic (0.03s)
        Error Trace:    image_test.go:76
	Error:      	Should be true
	Test:       	TestImageTransformBasic
FAIL
FAIL	github.com/gohugoio/hugo/resource	0.059s

Executing: go test github.com/gohugoio/hugo/resource -run "^TestImageTransformBasic$" || true
ok  	github.com/gohugoio/hugo/resource	(cached)
```

As for the test that broke earlier, that was actually a fault in the original test data. I fixed that and squashed it into the fix commit. My notes from the commit message:

>Fix image resize-then-fit test
>
>This test was written based on erroneous behaviour. We resize the image to 300x200px, and are now trying to fit it within a 50px square. The longest edge is 300 pixels, so we need to divide it by 6 (300 / 50 == 6). And then scale the shortest edge with the same proportion (200 / 6 == 33.33).
>
>The original test was transforming the original source image, hence the previous values:
>
>900 x 562
>900 / 50 == 18
>562 / 18 == 31.22

bep commented:
Thanks a lot for this fix, I really like ready-to-pull PRs.
